crates/circuit/src/interner.rs

```diff
@@ -10,79 +10,149 @@
 // copyright notice, and modified files need to carry a notice indicating
 // that they have been altered from the originals.
 
+use std::borrow::Borrow;
+use std::fmt;
 use std::hash::Hash;
-use std::sync::Arc;
+use std::marker::PhantomData;
 
-use hashbrown::HashMap;
-use pyo3::exceptions::PyRuntimeError;
-use pyo3::prelude::*;
+use indexmap::IndexSet;
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct Index(u32);
-
-impl IntoPy<PyObject> for Index {
-    fn into_py(self, py: Python<'_>) -> PyObject {
-        self.0.into_py(py)
-    }
+/// A key to retrieve a value (by reference) from an interner of the same type.  This is narrower
+/// than a true reference, at the cost that it is explicitly not lifetime bound to the interner it
+/// came from; it is up to the user to ensure that they never attempt to query an interner with a
+/// key from a different interner.
+#[derive(Debug, Eq, PartialEq)]
+pub struct Interned<T: ?Sized> {
+    index: u32,
+    // `Interned` is like a non-lifetime-bound reference to data stored in the interner.  Storing
+    // this adds a small amount more type safety to the interner keys when there's several interners
+    // in play close to each other.
+    _type: PhantomData<*const T>,
```

Comment:
Vibes mostly lol.  Logically, `Interned` doesn't _own_ the slice it refers to (e.g. when the `Interned` goes out of scope, the slice doesn't get dropped), it's just a way of retrieving the slice.  That makes it like a reference, except it doesn't have a statically checked lifetime, so `PhantomData<&'a T>` doesn't make sense.

From the perspective of the Rust compiler, there's no difference between me having written `_type: PhantomData<T>` and `_type: PhantomData<*const T>`, I just wrote the latter because that's closer to the actual erased type that `Interned` stores (and in a classical interner, the `Interned` type is often _literally_ a pointer).

Reply:
(strictly, I think `Send` would be auto-impl'ed if I'd used `PhantomData<T>` instead of `PhantomData<*const T>`, so maybe there's _technically_ a small difference)

Reply:
Hmm, alright, that's fine with me. I don't think we'll need to disturb this once it's in place anyhow, so it doesn't seem too important to explain the nuance with a comment etc.

Reply:
My intention had been that this bit of the comment:

> `Interned` is like a non-lifetime-bound reference to data stored in the interner.

was explaining the type of the `PhantomData`.  I'm fine to expand that code comment to be more in depth - I was already trying to answer this question in it anyway.

Reply:
Done in 2d56e17543.